File: ios/Runner/WXResponseHandler.m
//
//  WXResponseHandler.m
//  Runner
//
//  Created by litao on 2018/11/15.
//  Copyright © 2018年 The Chromium Authors. All rights reserved.
//

#import "WXResponseHandler.h"
#import "AppDelegate.h"
#import "StringUtil.h"
#import "WXKeys.h"

@implementation WXResponseHandler

const NSString *errStr = @"errStr";
const NSString *errCode = @"errCode";
const NSString *openId = @"openId";
const NSString *type = @"type";
const NSString *lang = @"lang";
const NSString *country = @"country";
const NSString *description = @"description";

+ (instancetype)defaultManager {
    static dispatch_once_t onceToken;
    static WXResponseHandler *instance;
    dispatch_once(&onceToken, ^{
      instance = [[WXResponseHandler alloc] init];
    });
    return instance;
}

FlutterMethodChannel *fluwexMethodChannel = nil;

- (void)setMethodChannel:(FlutterMethodChannel *)flutterMethodChannel {
    fluwexMethodChannel = flutterMethodChannel;
}

- (void)onResp:(BaseResp *)resp {
    NSLog(@"response");
    if ([resp isKindOfClass:[SendAuthResp class]]) {
        isWechatLoginStarted = NO;
        NSLog(@"auth response %@", resp.debugDescription);
        if (_delegate && [_delegate respondsToSelector:@selector(managerDidReceiveAuthResponse:)]) {
            SendAuthResp *authResp = (SendAuthResp *)resp;
            [_delegate managerDidReceiveAuthResponse:authResp];
        }
        SendAuthResp *authResp = (SendAuthResp *)resp;
        NSDictionary *result = @{
            description : authResp.description == nil ? @"" : authResp.description,
            errStr : authResp.errStr == nil ? @"" : authResp.errStr,
            errCode : @(authResp.errCode),
            type : authResp.type == nil ? @1 : @(authResp.type),
            country : authResp.country == nil ? @"" : authResp.country,
            lang : authResp.lang == nil ? @"" : authResp.lang,
            wxKeyPlatform : wxKeyIOS,
            @"code" : [StringUtil nilToEmpty:authResp.code],
            @"state" : [StringUtil nilToEmpty:authResp.state]
        };
        [fluwexMethodChannel invokeMethod:@"onAuthResponse" arguments:(id _Nullable)result];
    }
}

- (void)cancelLogin {
    NSLog(@"cancel login");
    NSMutableDictionary *args = [[NSMutableDictionary alloc] init];
    [args setValue:@"" forKey:@"description"];
    [args setValue:@"" forKey:@"errStr"];
    [args setValue:@(-2) forKey:@"errCode"];
    [args setValue:@(1) forKey:@"type"];
    [args setValue:@"" forKey:@"country"];
    [args setValue:@"" forKey:@"lang"];
    [args setValue:@"iOS" forKey:@"platform"];
    [args setValue:@"" forKey:@"code"];
    [args setValue:@"" forKey:@"state"];
    [fluwexMethodChannel invokeMethod:@"onAuthResponse" arguments:(id _Nullable)args];
}

@end

File: android/app/src/main/java/com/xiaoyan/ui/wechat/WXResponseHandler.java
package com.xiaoyan.ui.wechat;

import android.util.Log;

import com.tencent.mm.opensdk.modelbase.BaseResp;
import com.tencent.mm.opensdk.modelmsg.SendAuth;

import java.util.HashMap;

import io.flutter.plugin.common.MethodChannel;

public class WXResponseHandler {

    private static WXResponseHandler INSTANCE;

    private WXResponseHandler() {}

    public static WXResponseHandler getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new WXResponseHandler();
        }
        return INSTANCE;
    }

    private MethodChannel channel = null;

    private static final String errStr = "errStr";
    private static final String errCode = "errCode";
    private static final String openId = "openId";
    private static final String type = "type";

    public void setMethodChanner(MethodChannel channel) {
        this.channel = channel;
    }

    public void handleResponse(BaseResp response) {
        if (response instanceof SendAuth.Resp) {
            handleAuthResponse((SendAuth.Resp) response);
        }
    }

    private void handleAuthResponse(SendAuth.Resp response) {
        Log.e("WECHAT RESPONSE", "errCode: " + response.errCode);
        final HashMap<String, Object> map = new HashMap<>();
        map.put(WXKeys.PLATFORM, WXKeys.ANDROID);
        map.put(errCode, response.errCode);
        map.put("code", response.code);
        map.put("state", response.state);
        map.put("lang", response.lang);
        map.put("country", response.country);
        map.put(errStr, response.errStr);
        map.put(openId, response.openId);
        map.put("url", response.url);
        map.put(type, response.getType());
        map.put(WXKeys.TRANSACTION, response.transaction);
        channel.invokeMethod("onAuthResponse", map);
    }
}
